src/constructors.jl

```diff
@@ -0,0 +1,172 @@
+abstract type AbstractHMCSampler <: AbstractMCMC.AbstractSampler end
+
+##############
+### Custom ###
+##############
+
+"""
+    HMCSampler
+
+A `AbstractMCMC.AbstractSampler` for kernels in AdvancedHMC.jl.
+
+# Fields
+
+$(FIELDS)
+
+# Notes
+
+Note that all the fields have the prefix `initial_` to indicate
+that these will not necessarily correspond to the `kernel`, `metric`,
+and `adaptor` after sampling.
+
+To access the updated fields use the resulting [`HMCState`](@ref).
+"""
+Base.@kwdef struct HMCSampler{
+    K<:AbstractMCMCKernel,
+    M<:AbstractMetric,
+    A<:Adaptation.AbstractAdaptor,
+} <: AbstractHMCSampler
+    "[`AbstractMCMCKernel`](@ref)."
+    kernel::K
+    "[`AbstractMetric`](@ref)."
+    metric::M
+    "[`AbstractAdaptor`](@ref)."
+    adaptor::A
+    "Adaptation steps if any"
+    n_adapts::Int = 0
+end
+
+############
+### NUTS ###
+############
+"""
+    NUTS(n_adapts::Int, δ::Real; max_depth::Int=10, Δ_max::Real=1000, init_ϵ::Real=0)
+
+No-U-Turn Sampler (NUTS) sampler.
+
+# Fields
+
+$(FIELDS)
+
+# Usage:
+
+```julia
+NUTS(n_adapts=1000, δ=0.65)  # Use 1000 adaption steps, and target accept ratio 0.65.
+```
+"""
+struct NUTS{T<:AbstractFloat,I,D} <: AbstractHMCSampler
+    "`n_adapts::Int` : Number of adaptation steps."
+    n_adapts::Int
+    "`δ::Real` : Target acceptance rate for dual averaging."
+    δ::T
+    "`max_depth::Int` : Maximum doubling tree depth."
+    max_depth::Int
+    "`Δ_max::Real` : Maximum divergence during doubling tree."
+    Δ_max::T
+    "`init_ϵ::Real` : Initial step size; 0 means automatically searching using a heuristic procedure."
+    init_ϵ::T
+    "[`AbstractIntegrator`](@ref)."
+    integrator_method::I
+    "[`AbstractMetric`](@ref)."
+    metric_type::D
+end
+
+function NUTS(
```

Comment:
Is this constructor actually desired now that you switched to `@kwdef`? Or should we not use `@kwdef`?

Reply:
Hi! We no longer use `@kwdef`. That's why we got the constructor.

Reply:
We switched away from `@kwdef` because it doens't allow you to specify a default type parameter.

Reply:
`HMCSampler` still uses `@kwdef` it seems?